File: frontend/src/api/overviewMatrix.ts
/**
 * Overview matrix config API — ручная настройка квартальной матрицы «Сводного обзора»
 * по компании+году (выбор проектов, переопределения, свои пункты).
 */
import { api } from "./client";

export interface MatrixOverride {
  title?: string | null;
  due_date?: string | null;   // 'YYYY-MM-DD' или null
  quarter?: number | null;    // старт-квартал (0..3) или null = по дате
  quarter_end?: number | null; // конец-квартал (Гант-растяжка) или null = один квартал
  hidden?: boolean | null;
}

export interface MatrixCustomItem {
  id: string;
  direction_id?: string | null;
  direction_name?: string | null;
  title: string;
  due_date?: string | null;
  quarter?: number | null;
  quarter_end?: number | null;
}

export interface ManualProject {
  id: string;
  title: string;
  ref_project_id?: string | null;   // связанный системный проект (автоподстановка)
  quarter?: number | null;          // старт-квартал (0..3)
  quarter_end?: number | null;      // конец-квартал (Гант-растяжка)
  due_date?: string | null;
  details?: string | null;          // текст выноски (внизу отчёта)
}

export interface ManualDirection {
  id: string;
  name: string;
  projects: ManualProject[];
}

export interface MatrixConfig {
  hidden: string[];
  overrides: Record<string, MatrixOverride>;
  custom: MatrixCustomItem[];
  // Ручной отчёт: направления/проекты вписываются вручную, детали — в выноску.
  manual_directions: ManualDirection[];
}

export interface MatrixConfigResponse {
  company_id: string;
  year: number;
  config: MatrixConfig;
  updated_at?: string | null;
  updated_by_name?: string | null;
}

export function emptyMatrixConfig(): MatrixConfig {
  return { hidden: [], overrides: {}, custom: [], manual_directions: [] };
}

export const overviewMatrixApi = {
  async get(companyId: string, year: number): Promise<MatrixConfigResponse> {
    const r = await api.get<MatrixConfigResponse>(`/overview-matrix/${companyId}/${year}`);
    return r.data;
  },
  async save(companyId: string, year: number, config: MatrixConfig): Promise<MatrixConfigResponse> {
    const r = await api.put<MatrixConfigResponse>(`/overview-matrix/${companyId}/${year}`, config);
    return r.data;
  },
};

File: backend/app/schemas/overview_matrix.py
"""Overview matrix config schemas — настройка квартальной матрицы Сводного обзора."""
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class MatrixOverride(BaseModel):
    """Переопределение отображения реального проекта (всё опционально)."""
    title: Optional[str] = Field(None, max_length=512)
    due_date: Optional[str] = None      # ISO-дата 'YYYY-MM-DD' или null
    quarter: Optional[int] = Field(None, ge=0, le=3)      # старт-квартал; null = по дате
    quarter_end: Optional[int] = Field(None, ge=0, le=3)  # конец-квартал (Гант-растяжка); null = один квартал
    hidden: Optional[bool] = None        # на случай хранения скрытия в overrides


class MatrixCustomItem(BaseModel):
    """Произвольный пункт (которого нет в системе)."""
    id: str
    direction_id: Optional[str] = None
    direction_name: Optional[str] = Field(None, max_length=255)
    title: str = Field(..., max_length=512)
    due_date: Optional[str] = None
    quarter: Optional[int] = Field(None, ge=0, le=3)
    quarter_end: Optional[int] = Field(None, ge=0, le=3)


class MatrixManualProject(BaseModel):
    """Проект ручного отчёта: название (вписывают), квартал(ы), дата, детали для выноски."""
    id: str
    title: str = Field("", max_length=512)
    ref_project_id: Optional[str] = None                  # связанный системный проект (автоподстановка)
    quarter: Optional[int] = Field(None, ge=0, le=3)      # старт-квартал
    quarter_end: Optional[int] = Field(None, ge=0, le=3)  # конец-квартал (Гант-растяжка)
    due_date: Optional[str] = None
    details: Optional[str] = Field(None, max_length=4000)  # текст выноски (внизу отчёта)


class MatrixManualDirection(BaseModel):
    """Направление ручного отчёта (строка матрицы) — название вписывают вручную."""
    id: str
    name: str = Field("", max_length=255)
    projects: list[MatrixManualProject] = Field(default_factory=list)


class MatrixConfig(BaseModel):
    """Полный конфиг матрицы для (company, year)."""
    model_config = ConfigDict(extra="ignore")

    hidden: list[str] = Field(default_factory=list)            # скрытые project_id
    overrides: dict[str, MatrixOverride] = Field(default_factory=dict)
    custom: list[MatrixCustomItem] = Field(default_factory=list)
    # Ручной отчёт (новый режим): направления и проекты вписываются вручную,
    # детали проектов выносятся в сноску внизу отчёта. Если непусто — отчёт
    # рендерится из этого, а не из системных проектов.
    manual_directions: list[MatrixManualDirection] = Field(default_factory=list)


class MatrixConfigResponse(BaseModel):
    company_id: UUID
    year: int
    config: MatrixConfig
    updated_at: Optional[datetime] = None
    updated_by_name: Optional[str] = None
